clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -455,7 +456,13 @@ private static int parseAcks(String acksString) {
                         " to class " + producerConfig.getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
                         " specified in value.serializer");
             }
-            int partition = partition(record, serializedKey, serializedValue, metadata.fetch());
+            // If topic is deleted between the start of this method when metadata was obtained
+            // and now, fail the send request.
+            Cluster cluster = metadata.fetch();
```

Comment:
@ewencp Thank you for the review. I have updated the code to hold on to `Cluster` after a metadata update to avoid small timing windows like this issue in methods where metadata may have changed. For RecordAccumulator, it is better to get the latest metadata since a significant amount of time could have passed between accumulating and sending. Changes in the cluster are already handled properly in this case. Invalid partitions are already handled in the producer which covers changes to partitions, and since partitions cannot be deleted in Kafka, there is less scope for issues  compared to topic deletion.